Add support for reproducible builds

adamws opened the pull request:
See [1] for more info about reproducible builds. In case of zip files we need to adjust timestamps. After this change, each built zip archive will have same hash (as long as packaged files do not change)

[1] https://hatch.pypa.io/latest/config/build/#reproducible-builds

adamws committed:
Add support for reproducible builds

See [1] for more info about reproducible builds. In case of zip files we
need to adjust timestamps. After this change, each built zip archive
will have same hash (as long as packaged files do not change)

[1] https://hatch.pypa.io/latest/config/build/#reproducible-builds

dairiki commented on hatch_zipped_directory/builder.py:
```diff
@@ -24,26 +26,43 @@
         self.root_path = PurePosixPath(root_path)
         self.zipfd = zipfd
+        self.reproducible = reproducible
+        self.timestamp: int | None = (
```
Is the `timestamp` attribute necessary (or can it just be local variable)?

adamws replied:
only local variable

dairiki commented on hatch_zipped_directory/builder.py:
```diff
@@ -24,26 +26,43 @@
+        if self.ziptime:
+            info.date_time = self.ziptime
+        with open(included_file.path, "rb") as f:
+            self.zipfd.writestr(info, f.read())
```
I'm not crazy about the use of `ZipFile.writestr` here.  I'd like to avoid reading the whole file into memory, in case of large files.

One alternative would be to use `ZipFile.open` (which will accept a `ZipInfo` argument) and `shutil.copyfileobj` directly, but, sadly, it appears that would require duplicating a bunch of special-case/metadata logic (e.g. to handle directories, and to set `compress_type`) from [ZipFile.write](https://github.com/python/cpython/blob/f2584eade378910b9ea18072bb1dab3dd58e23bb/Lib/zipfile/__init__.py#L1805-L1836). 
(Also compare to hatchling's  [WheelArchive.add_file](https://github.com/pypa/hatch/blob/7dac9856d2545393f7dd96d31fc8620dde0dc12d/backend/src/hatchling/builders/wheel.py#L84-L112).)
(I'm not sure, but it looks as if maybe `add_file` never gets called for directories?  If so, that would make this less messy.)

adamws replied:
I honestly did not consider large files, kicad plugins are size limited so that's not real concern for me. But I agree that hatch-zipped-directory should use better approach.

dairiki commented on hatch_zipped_directory/builder.py:
```diff
@@ -24,26 +26,43 @@
     @contextmanager
-    def open(cls, dst: str | os.PathLike[str], root_path: str) -> Iterator[ZipArchive]:
+    def open(
+        cls, dst: str | os.PathLike[str], root_path: str, *, reproducible: bool
```
Add a default value for `reproducible`?

dairiki commented on hatch_zipped_directory/builder.py:
```diff
@@ -24,26 +26,43 @@
 
 class ZipArchive:
-    def __init__(self, zipfd: ZipFile, root_path: str):
+    def __init__(self, zipfd: ZipFile, root_path: str, *, reproducible: bool):
```
Add a default value for `reproducible`?

adamws replied:
I made it keyword-only and explicit to be https://beta.ruff.rs/docs/rules/boolean-type-hint-positional-argument/ compatible

dairiki replied:
I have no issue with making it keyword-only.  I would favor adding a default value, i.e.

```py
class ZipArchive:
    def __init__(self, zipfd: ZipFile, root_path: str, *, reproducible: bool = False):
        ...
```


This is admittedly a picky stylistic nit.

It is true that our code always provides a value for the argument, so it's not critical that we provide a default.  But I do view the `reproducible` option as "optional" behavior that is not central to the operation of `ZipArchive`.  Adding a default value expresses that notion. (It also prevents breaking any external code — though I suspect there is none — that may use `ZipArchive` directly.)

I don't have any strong preference as to whether the default should be `reproducible=False` or `reproducible=True`.  Defaulting to `False` maintains existing behavior, but, since hatch defaults to `reproducible=True`, that would be a reasonable default as well.

dairiki commented on hatch_zipped_directory/builder.py:
```diff
@@ -24,26 +26,43 @@
+        self.timestamp: int | None = (
+            get_reproducible_timestamp() if reproducible else None
+        )
+        self.ziptime: ZipTime | None = (
```
Nit: perhaps we can find a better name for `ziptime`.  Maybe `reproducible_timestamp` or `reproducible_date_tuple`?  Some name that indicates its purpose (reproducibility), in any case.

dairiki commented on tests/test_builder.py:
```diff
@@ -11,6 +12,16 @@
+def getsha256(filename) -> str:
+    sha256 = hashlib.sha256()
+    with open(filename, "rb") as f:
+        while data := f.read(READ_SIZE):
```
The Walrus Operator (`:=`) is not supported in python 3.7 (so its use breaks our CI tests.)

I suppose we could drop support for python 3.7, but I don't see the point of doing so just for this one case.

adamws replied:
without `:=` used

```python
def getsha256(filename) -> str:
    hash = hashlib.sha256()
    with io.open(filename, "rb") as f:
        data = f.read(READ_SIZE)
        while data:
            hash.update(data)
            data = f.read(READ_SIZE)
    return hash.hexdigest()
```

dairiki commented on hatch_zipped_directory/builder.py:
```diff
@@ -2,20 +2,22 @@
+from typing import Tuple
 from zipfile import ZIP_DEFLATED
-from zipfile import ZipFile
+from zipfile import ZipFile, ZipInfo
```
Style: [reorder-python-imports](https://github.com/asottile/reorder_python_imports) (which is run by our [.pre-commit-config.yaml](https://github.com/dairiki/hatch-zipped-directory/blob/master/.pre-commit-config.yaml)) was each import on a separate line.

If you use [pre-commit](https://pre-commit.com/), this will be fixed automatically on git commit.

dairiki commented on tests/test_builder.py:
```diff
@@ -21,66 +32,95 @@ def zip_contents(path):
-    assert len(list(tmp_path.iterdir())) == 0
+def assert_zip_timestamps(archive_path, reproducible: bool):
+    if reproducible:
+        condition = lambda t: t == (2020, 2, 2, 0, 0, 0)
```
Flake8 would prefer the use of defs rather than lambdas here ([E731](https://www.flake8rules.com/rules/E731.html)).
(https://www.flake8rules.com/rules/E731.html)).

Perhaps
```py
def is_reproducible_ts(ts: ZipTime) -> bool:
    return ts == (2020, 2, 2, 0, 0, 0)

# (inline zip_timestamps(), since it is only used here)
with ZipFile(archive_path) as zf:
    if reproducible:
        assert all(is_reproducible_ts(info.date_time) for info in zf.infolist())
    else:
        assert not any(is_reproducible_ts(info.date_time) for info in zf.infolist())
```


----

As an alternative take:

In general, we care about testing two different things, depending on whether we are creating reproducible archives are not.

- In the reproducible case, we care that all the timestamps are crowbarred to SOURCE_DATE_EPOCH.

- In the non-reproducible case, we care that timestamps for files in the zip archive match those of the source files on the filesystem.

Perhaps, instead of parametrizing the tests on `reproducible`, just add a couple of new test methods to `TestZipArchive` that test for those behaviors.  (E.g. `TestZipArchive.test_reproducible_timestamps` and `TestZipArchive.test_copies_timestamps`.)

Parametrizing `reproducible` has some value (in order to check that all the other tests pass for either value).  But, even if that parameterization is kept — in the spirit of "each test should test just one thing" — I'd vote for adding new tests to test the zip timestamps are correct.

dairiki commented on tests/test_builder.py:
```diff
@@ -179,10 +219,41 @@ def test_ZippedDirectoryBuilder_build(builder, project_root, tmp_path, arcname_p
+        {"reproducible": False},
+    ],
+)
+def test_ZippedDirectoryBuilder_reproducible(
```
It might be worth splitting this into two tests — one for a reproducible build, and one for a non-reproducible build — rather than the single parametrized test.

As noted in a previous comment, we care about subtly different things in the two cases.
- In the reproducible case, we care (as tested here) that final artifact does not depend on the timestamps of the constituent input files.
- In the non-reproducible case, we care that the files in the archive get the correct timestamps (i.e. that the timestamps in the zip archive matches those of the input files.)  (To test this we only need to build once.)

adamws replied:
I agree. I took shortest route.

dairiki reviewed (changes requested):
@adamws Thank you for the PR!  I think adding the capability of reproducible builds is a valuable addition.

Currently, this code fails our CI tests, due to various style nits.

Here are some comments which call out those nits, and add a few more.

The most important has to do with not reading whole files into memory in `ZipArchive.add_file`.

(If you'd rather I fix these, say so.  I don't want to mess with your code, however, if you feel like addressing these issues.)

adamws commented:
> @adamws Thank you for the PR! I think adding the capability of reproducible builds is a valuable addition.
> 
> Currently, this code fails our CI tests, due to various style nits.
> 
> Here are some comments which call out those nits, and add a few more.
> 
> The most important has to do with not reading whole files into memory in `ZipArchive.add_file`.
> 
> (If you'd rather I fix these, say so. I don't want to mess with your code, however, if you feel like addressing these issues.)

Hello, this code is derived from my project: https://github.com/adamws/hatch-kicad
Looking at `hatch-zipped-directory` code helped me to understand how to write hatch plugins so I wanted to give something back. You are welcome to take this patch and modify it.

adamws committed:
Fix python 3.7 compatibility

adamws committed:
Fix import order

dairiki committed:
fixup! Fix import order

dairiki committed:
refactor(tests): add separate tests for reproducible builds

dairiki committed:
fix(ZipArchive.write_file): restore setting timestamp to now

Restore the behavior (from ZipFile.writestr) of setting the timestamp
to the current time for non-reproducible builds.

dairiki committed:
refactor(ZipArchive.add_file): do not read large files into memory

Plus a couple of stylistic refactorings

dairiki committed:
fix(ZipArchive.add_file): normalize file mode bits in reproducible mode

Ripped from hatchling's WheelArchive.add_file:

Popular VCSs only track whether a file is executable or not. The exact
permissions can vary on systems with different umasks. Normalizing to
644 (non executable) or 755 (executable) makes builds more
reproducible.

dairiki commented:
> You are welcome to take this patch and modify it.

@adamws  Thank you.  I think I've addressed the issues raised in my comments.

Again, thank you for the PR.  I think it's a good addition.

dairiki committed:
docs: update README and CHANGES

dairiki commented:
Released in [0.1.0b4](https://github.com/dairiki/hatch-zipped-directory/releases/tag/v0.1.0b4).
